webrender/src/device/gfx/buffer.rs

```diff
@@ -268,93 +272,120 @@ impl<B: hal::Backend> InstancePoolBuffer<B> {
+        free_buffers: &mut SmallVec<[InstancePoolBuffer<B>; 16]>,
+    ) -> std::ops::Range<usize> {
+        fn instance_data_to_u8_slice<T: Copy>(data: &[T]) -> &[u8] {
+            unsafe {
+                std::slice::from_raw_parts(
+                    data.as_ptr() as *const u8,
+                    data.len() * mem::size_of::<T>(),
+                )
+            }
+        }
+
+        let data_stride = mem::size_of::<T>();
+        let mut range = 0..0;
+        let mut first_iteration = true;
+        while !instance_data.is_empty() {
+            let need_new_buffer = self.buffers.is_empty()
+                || !self.current_buffer().can_store_data(data_stride);
+            if need_new_buffer {
+                let buffer = match free_buffers.pop() {
+                    Some(b) => b,
+                    None => InstancePoolBuffer::new(
                         device,
                         heaps,
                         hal::buffer::Usage::VERTEX,
-                        self.data_stride,
                         self.alignment_mask,
                         self.non_coherent_atom_size_mask,
-                    ))
-                }
-            }
-
-            let update_size = if (self.current_buffer().offset + data.len()) * self.data_stride
-                > self.current_buffer().buffer.buffer_size
-            {
-                self.current_buffer().buffer.buffer_size / self.data_stride
-                    - self.current_buffer().offset
+                    ),
+                };
+                self.buffers.push(buffer);
+                self.next_buffer_index += 1;
             } else {
-                data.len()
-            };
-
-            self.buffers[self.current_buffer_index].update(device, &data[0 .. update_size]);
-
-            data = &data[update_size ..]
+                self.buffers[self.next_buffer_index - 1].align_offset_to(data_stride);
+            }
+            if first_iteration {
+                range.start = self.next_buffer_index - 1;
+                first_iteration = false;
+            }
+            let update_size = (self.current_buffer().free() / data_stride).min(instance_data.len());
+            self.buffers[self.next_buffer_index - 1].update(device, instance_data_to_u8_slice(&instance_data[0 .. update_size]), data_stride);
+            instance_data = &instance_data[update_size ..];
         }
+        range.end = self.next_buffer_index;
+        range
     }
 
     fn current_buffer(&self) -> &InstancePoolBuffer<B> {
-        &self.buffers[self.current_buffer_index]
+        &self.buffers[self.next_buffer_index - 1]
     }
 
-    pub(super) fn reset(&mut self) {
-        for buffer in &mut self.buffers {
+    pub(super) fn reset(&mut self, free_buffers: &mut SmallVec<[InstancePoolBuffer<B>; 16]>) {
+        // Keep one buffer and move the others back to the free set pool.
+        while self.buffers.len() > 1 {
+            let mut buffer = self.buffers.pop().unwrap();
```

Comment:
`for mut buffer in self.buffers.drain(1 .. ) {..}`

Reply:
That's definitely better, thanks!